
```{r}
message(" ## Loading libraries: dplyr, readr, coloc, GenomicRanges, Rsamtools, optparse")
suppressPackageStartupMessages(library("dplyr"))
suppressPackageStartupMessages(library("readr"))
suppressPackageStartupMessages(library("coloc"))
suppressPackageStartupMessages(library("GenomicRanges"))
suppressPackageStartupMessages(library("Rsamtools"))
suppressPackageStartupMessages(library("optparse"))
library(plotly)
library(reshape2)
library(viridis)
library(hrbrthemes)
library(ggrepel)
library(ggpubr)

save_ggplots <- function(plot, path = ".", filename = "unnamed_plot", height = 15, width = 15){
  if (!dir.exists(path)) {
    dir.create(path = path, recursive = TRUE)
  }

ggsave(plot = plot,
       filename = paste0(filename, "_", width, "x", height, ".pdf"), 
       path = path,
       device = "pdf", 
       height = height, 
       width = width,
       units = "cm",
       dpi = 300)
}

read_colocs_to_df <- function(file_dir, ...) {
  file_list <- list.files(file_dir, recursive = TRUE ,full.names = TRUE)
  dataset = data.frame()
  
  for (file in file_list){
    print(file)    
  
    # if the merged dataset does exist, append to it
    temp_dataset <-read_tsv(file, trim_ws = TRUE)
    temp_dataset <- temp_dataset %>% dplyr::filter(...)
    dataset<-rbind(dataset, temp_dataset)
  }
  return(dataset)
}

output_dir = paste0("../plots/", format(Sys.time(), "%d_%b_%Y"))
if (!dir.exists(paste0(output_dir, "/plotly"))) {
  dir.create(paste0(output_dir, "/plotly"), recursive = TRUE)
}
```

First we read necessary data and metadata files and do the preliminary joining operations
```{r}
# # read all the necessary coloc data
# GTExV8_sign_colocs <- read_colocs_to_df(file_dir = "../data/coloc_results/results_coloc_GTExV8_15Sept2020/", PP.H4.abf>=0.8)
# eqtlCat_rnaseq_sign_colocs <- read_colocs_to_df(file_dir = "../data/coloc_results/results_coloc_eQTLCatR3_rnaseq_16Sept2020/", PP.H4.abf>=0.8)
# eqtlCat_ma_sign_colocs <- read_colocs_to_df(file_dir = "../data/coloc_results/results_coloc_eQTLCatR3_microarray_15Sept2020/", PP.H4.abf>=0.8)
# 
# # replace probeIDs with geneIDs in microarray colocs
# ma_pheno_meta <- read_tsv("../data/phenotype_metadata/HumanHT-12_V4_Ensembl_96_phenotype_metadata.tsv.gz") %>% 
#   select(phenotype_id, gene_id)
# colnames(ma_pheno_meta) <- c("molecular_trait_id", "gene_id")
# 
# eqtlCat_ma_sign_colocs_replaced <- eqtlCat_ma_sign_colocs %>% 
#   left_join(ma_pheno_meta) %>% 
#   mutate(molecular_trait_id = gene_id) %>% 
#   select(-gene_id)
# 
# # row binf rnaseq and microarray colocs into one df
# eqtlCat_all_sign_colocs <- rbind(eqtlCat_rnaseq_sign_colocs, eqtlCat_ma_sign_colocs_replaced)
# eqtlCat_all_sign_colocs <- eqtlCat_all_sign_colocs %>% filter(gwas_id != "RA-ieu-a-833")

# read other metadata files needed for analysis
qtl_group_samplesizes <- read_tsv("data/significant_colocs/qtl_group_samplesizes.tsv")
ld_blocks <-  read_tsv("data/significant_colocs/ind_regions_Hg38.bed")

# write_tsv(GTExV8_sign_colocs, "../data/TSVs/colocs_merged_R3/GTExV8_sign_colocs.tsv")
# write_tsv(eqtlCat_all_sign_colocs, "../data/TSVs/colocs_merged_R3/eqtlCat_all_sign_colocs.tsv")
```

```{r}

GTExV8_sign_colocs <- read_tsv("data/significant_colocs/GTExV8_sign_colocs.tsv")
eqtlCat_all_sign_colocs <- read_tsv("data/significant_colocs/eqtlCat_all_sign_colocs.tsv")
```


```{r}
gwas_ids <- GTExV8_sign_colocs$gwas_id %>% unique()

distinct_eqtlcat_egenes <- list()
distinct_gtex_egenes <- list()
union_egenes_gwas <- list()
uniq_eqtlcat_egenes_gwas <- list()

df_novel_eqtlcat_colocs <- data.frame()
for (gwasid in gwas_ids) {
  eqtl_cat_colocs_by_gwas = eqtlCat_all_sign_colocs %>% filter(gwas_id == gwasid)
  uniq_eqtlcat_egenes <- eqtl_cat_colocs_by_gwas %>% pull(molecular_trait_id) %>% unique() 
  uniq_gtex_egenes <- GTExV8_sign_colocs %>% filter(gwas_id == gwasid) %>% pull(molecular_trait_id) %>% unique() 
  
  common_egenes <- intersect(uniq_eqtlcat_egenes, uniq_gtex_egenes)
  distinct_eqtlcat_egenes[[gwasid]] <- uniq_eqtlcat_egenes[!uniq_eqtlcat_egenes %in% common_egenes]
  distinct_gtex_egenes[[gwasid]] <- uniq_gtex_egenes[!uniq_gtex_egenes %in% common_egenes]
  union_egenes_gwas[[gwasid]] <- union(uniq_eqtlcat_egenes, uniq_gtex_egenes)
  
  eqtl_cat_distinct_colocs_by_gwas <- eqtl_cat_colocs_by_gwas %>% filter(molecular_trait_id %in% distinct_eqtlcat_egenes[[gwasid]])
  df_novel_eqtlcat_colocs <- rbind(df_novel_eqtlcat_colocs, eqtl_cat_distinct_colocs_by_gwas)
}
# write_tsv(df_novel_eqtlcat_colocs, "../data/TSVs/colocs_merged_R3/novel/novel_eqtlcat_colocs_by_mtid.tsv")
```

```{r}
distinct_eqtlcat_ld_blocks <- list()
distinct_gtex_ld_blocks <- list()
union_ld_blocks <- list()

novel_ld_block_colocs <- data.frame()

for (gwasid in gwas_ids) {
  eqtl_cat_novel_colocs_by_gwas <- df_novel_eqtlcat_colocs %>% filter(gwas_id == gwasid)
  gtex_colocs_by_gwas <- GTExV8_sign_colocs %>% filter(gwas_id == gwasid)
  
  # merge according to chromosome, start and end positions
  gtex_merged = sqldf::sqldf("
    SELECT *
    FROM gtex_colocs_by_gwas d1 JOIN ld_blocks d2
    ON d1.chromosome = d2.chr
    AND d1.position >= d2.start
    AND d1.position <= d2.end")
  
  gtex_unique_ld_blocks <- gtex_merged %>% pull(ID_hg38) %>% unique()
  
  eqtlcat_merged = sqldf::sqldf("
    SELECT *
    FROM eqtl_cat_novel_colocs_by_gwas d1 JOIN ld_blocks d2
    ON d1.chromosome = d2.chr
    AND d1.position >= d2.start
    AND d1.position <= d2.end")
  
  eqtlcat_unique_ld_blocks <- eqtlcat_merged %>% pull(ID_hg38) %>% unique()
    
  common_ld_blocks <- intersect(gtex_unique_ld_blocks, eqtlcat_unique_ld_blocks)
  distinct_eqtlcat_ld_blocks[[gwasid]] <- eqtlcat_unique_ld_blocks[!eqtlcat_unique_ld_blocks %in% common_ld_blocks]
  distinct_gtex_ld_blocks[[gwasid]] <- gtex_unique_ld_blocks[!gtex_unique_ld_blocks %in% common_ld_blocks]
  union_ld_blocks[[gwasid]] <- union(eqtlcat_unique_ld_blocks, gtex_unique_ld_blocks)
  
  colocs_in_novel_ld_blocks <- eqtlcat_merged %>% 
    filter(ID_hg38 %in% distinct_eqtlcat_ld_blocks[[gwasid]]) %>%
    select(-chr,-start,-end, -ID_hg19)
  
  novel_ld_block_colocs <- novel_ld_block_colocs %>% rbind(colocs_in_novel_ld_blocks)
}

# write_tsv(novel_ld_block_colocs, "../data/TSVs/colocs_merged_R3/novel/novel_eqtlcat_colocs_in_novel_ldblocks.tsv")
```

```{r}
novel_ld_block_colocs$gwas_id %>% unique()
novel_ld_block_colocs$qtl_subset %>% unique()

novel_ld_block_colocs <- read_tsv("data/significant_colocs/novel_eqtlcat_colocs_in_novel_ldblocks_relative_to_GTExV8.tsv")

novel_ld_block_colocs %>% 
  filter(gwas_id == "LC-ebi-a-GCST004627") %>% 
  filter(qtl_subset == "BLUEPRINT_PE.T-cell_ge") %>% 
  pull(ID_hg38) %>% 
  unique()
```

```{r}
novel_ld_block_based_counts <- novel_ld_block_colocs %>% group_by(gwas_id, qtl_subset) %>% summarise(
  n_colocs_in_novel_ld_blocks = n(), 
  n_unique_novel_ld_blocks = length(unique(ID_hg38)))

## HEIGHT
gwas_colocs = novel_ld_block_based_counts %>% filter(gwas_id == "HEIGHT-ukb-a-389")

zero_egene_qtlSubsets <- qtl_group_samplesizes[!qtl_group_samplesizes$qtl_subset %in% gwas_colocs$qtl_subset,] %>%
    select(-sample_size, -novel_egene_count) %>% 
    mutate(gwas_id = gwasid, n_colocs_in_novel_ld_blocks = 0, n_unique_novel_ld_blocks = 0)

gwas_colocs <- dplyr::bind_rows(gwas_colocs, zero_egene_qtlSubsets) %>% 
    left_join(qtl_group_samplesizes %>% select(qtl_subset, sample_size))

top <- gwas_colocs %>% arrange(-n_unique_novel_ld_blocks) %>% pull(n_unique_novel_ld_blocks)
  
  # make scatter plots with repel labels
  temp_plot <- ggplot(gwas_colocs, aes(x = sample_size, y = n_unique_novel_ld_blocks, label=qtl_subset)) +
    geom_vline(xintercept = 150, color = "darkred", alpha = 0.5, size = 2) +
    geom_point(size = 2, alpha=0.8) + theme_light() +
    geom_label_repel(
      data = subset(gwas_colocs, n_unique_novel_ld_blocks >= mean(unique(top)[2]) & n_unique_novel_ld_blocks>0),
      aes(label = qtl_subset),
      size = 2,
      box.padding = unit(0.35, "lines"),
      point.padding = unit(0.3, "lines")
    ) +
    ggplot2::labs(x="Dataset sample size", 
                  y="Count of unique novel LD blocks from dataset", 
                  title = paste0("Plot of the ", gwasid, " novel LD blocks counts\nwhen using eQTLCatalogue relative to GTExV8"),
                  color = "Cell type class") + 
    theme(plot.title = element_text(size = 10))


```

```{r}
# Make scatter plots per GWAS (sample size in X, novel ld block counts in Y)
qtlsubset_classes <- read_tsv("../data/TSVs/coloc_temp_docs/qtl_subset_classes.tsv")

novel_ld_block_based_counts <- novel_ld_block_colocs %>% group_by(gwas_id, qtl_subset) %>% summarise(
  n_colocs_in_novel_ld_blocks = n(), 
  n_unique_novel_ld_blocks = length(unique(ID_hg38))) 

for (gwasid in gwas_ids) {
  gwas_colocs = novel_ld_block_based_counts %>% filter(gwas_id == gwasid)
  zero_egene_qtlSubsets <- qtl_group_samplesizes[!qtl_group_samplesizes$qtl_subset %in% gwas_colocs$qtl_subset,] %>%
    select(-sample_size, -novel_egene_count) %>% 
    mutate(gwas_id = gwasid, n_colocs_in_novel_ld_blocks = 0, n_unique_novel_ld_blocks = 0) 

  gwas_colocs <- gwas_colocs %>% rbind(zero_egene_qtlSubsets) %>% 
    left_join(qtl_group_samplesizes %>% select(qtl_subset, sample_size)) %>% 
    left_join(qtlsubset_classes %>% select(qtl_subset, cell_type_class))
  
  top <- gwas_colocs %>% arrange(-n_unique_novel_ld_blocks) %>% pull(n_unique_novel_ld_blocks)
  
  # make scatter plots with repel labels
  temp_plot <- ggplot(gwas_colocs, aes(x = sample_size, y = n_unique_novel_ld_blocks, label=qtl_subset)) +
    # geom_rect(data = gwas_colocs, mapping=aes(xmin=0, xmax=150, ymin=0,ymax=Inf), fill = "darkred", alpha=0.2) +
    geom_vline(xintercept = 150, color = "darkred", alpha = 0.5, size = 2) +
    geom_point(aes(color = factor(cell_type_class)), size = 2, alpha=0.8) + theme_light() +
    geom_label_repel(
      data = subset(gwas_colocs, n_unique_novel_ld_blocks >= mean(unique(top)[2]) & n_unique_novel_ld_blocks>0),
      aes(label = qtl_subset),
      size = 2,
      box.padding = unit(0.35, "lines"),
      point.padding = unit(0.3, "lines")
    ) +
    ggplot2::labs(x="Dataset sample size", 
                  y="Count of unique novel LD blocks from dataset", 
                  title = paste0("Plot of the ", gwasid, " novel LD blocks counts\nwhen using eQTLCatalogue relative to GTExV8"),
                  color = "Cell type class") + 
    theme(plot.title = element_text(size = 10))
  
  save_ggplots(plot = temp_plot,
               filename = paste0(gwasid, ".novel_ld_blocks"),
               path = paste0(output_dir, "/scatter_plots_LD_blocks/"), 
               width = 16, height = 12)
  
  # make scatter plots without repel labels for plotly
  temp_plot <- ggplot(gwas_colocs, aes(x = sample_size, y = n_unique_novel_ld_blocks, label=qtl_subset)) +
    geom_point(aes(color = factor(qtl_subset)), show.legend = FALSE) + theme_light() +
    ggtitle(paste0("Plot of the ", gwasid, " novel LD blocks counts when using eQTLCatalogue relative to GTExV8")) +
    theme(plot.title = element_text(size = 10))
  
  ggplotly_plot <- plotly::ggplotly(temp_plot)
  htmlwidgets::saveWidget(widget = plotly::as_widget(ggplotly_plot),
                          file = file.path(normalizePath(paste0(output_dir, "/plotly")),
                                           paste0(gwasid,"_novel_ld_blocks_scatter.html")),
                          libdir = "dependencies")
}

```


```{r}
# get proportions of novel LD blocks in comparison to GTExV8 

novel_ld_block_based_counts_prop <- novel_ld_block_colocs %>% group_by(gwas_id) %>% summarise(
  n_colocs_in_novel_ld_blocks = n(), 
  n_unique_novel_ld_blocks = length(unique(ID_hg38)))

union_ld_block_counts <- as.data.frame(lengths(union_ld_blocks))
union_ld_block_counts$gwas_id = rownames(union_ld_block_counts)
colnames(union_ld_block_counts) <- c("n_union_ld_blocks", "gwas_id")

distinct_gtex_ld_blocks_counts <- as.data.frame(lengths(distinct_gtex_ld_blocks))
distinct_gtex_ld_blocks_counts$gwas_id = rownames(distinct_gtex_ld_blocks_counts)
colnames(distinct_gtex_ld_blocks_counts) <- c("n_distinct_gtex_ld_blocks", "gwas_id")

novel_ld_block_counts = novel_ld_block_based_counts_prop %>% 
  left_join(distinct_gtex_ld_blocks_counts) %>% 
  left_join(union_ld_block_counts) %>% 
  mutate(n_common_ld_blocks = n_union_ld_blocks - n_distinct_gtex_ld_blocks - n_unique_novel_ld_blocks, 
         novel_ld_block_perc = paste0(round(100*n_unique_novel_ld_blocks/n_union_ld_blocks, digits = 2), "%"))

# write_tsv(novel_ld_block_counts, "../data/TSVs/colocs_merged_R3/novel/novel_ld_blocks_counts.tsv")

novel_ld_block_counts_melted <- melt(novel_ld_block_counts %>% select(-n_colocs_in_novel_ld_blocks, -n_union_ld_blocks),
                                     id.vars=c("gwas_id","novel_ld_block_perc"))

plot_distinct_counts <- ggplot(novel_ld_block_counts_melted, aes(fill=variable, y=value, x=gwas_id, label = novel_ld_block_perc)) + 
    geom_bar(position="stack", stat="identity") +
    scale_fill_viridis(discrete = T) +
    theme_light() +
    ggplot2::labs(x="GWAS ID", 
                  y="# of independent LD blocks", 
                  title = "Proportion of novel independent LD blocks") + 
    theme(axis.text.x = element_text(angle = 90), ) + scale_y_continuous(limits = c(0,800))


ggplotly_plot <- plotly::ggplotly(plot_distinct_counts)
htmlwidgets::saveWidget(widget = plotly::as_widget(ggplotly_plot),
                        file = file.path(normalizePath(paste0(output_dir, "/plotly")), "novel_ld_blocks_proportions_new.html"),
                        libdir = "dependencies")

novel_ld_block_counts_melted$novel_ld_block_perc[novel_ld_block_counts_melted$variable != "n_unique_novel_ld_blocks"] <- NA
plot_distinct_counts <- ggplot(novel_ld_block_counts_melted, aes(fill=variable, y=value, x=gwas_id, label = novel_ld_block_perc)) + 
    geom_bar(position="stack", stat="identity") +
    geom_text(aes(label=novel_ld_block_perc), vjust = -1, size = 2, position = "stack") +
    scale_fill_viridis(discrete = T) +
    theme_light() +
    ggplot2::labs(x="GWAS ID", 
                  y="# of independent LD blocks", 
                  title = "Proportion of novel independent LD blocks") + 
    theme(axis.text.x = element_text(angle = 90), ) + scale_y_continuous(limits = c(0,800))

save_ggplots(plot = plot_distinct_counts, path = paste0(output_dir, "/barplots"), filename = "novel_ld_blocks_proportions_new" , width = 20)
```

```{r}
# violin plots generation
gwas_classes <- read_tsv("../data/TSVs/colocs_merged_R3/gwas_classes.tsv")
qtlsubset_classes <- read_tsv("../data/TSVs/coloc_temp_docs/qtl_subset_classes.tsv")

novel_ld_block_based_counts_byqtl <- novel_ld_block_based_counts %>% 
  left_join(qtlsubset_classes %>% select(-n_colocs_in_novel_ld_blocks, -n_unique_novel_ld_blocks)) %>% 
  mutate(sample_size_norm_novel_LD_block_count = n_unique_novel_ld_blocks/sample_size) %>% 
  left_join(gwas_classes)

gwas_ids <- novel_ld_block_based_counts_byqtl %>% pull(gwas_id) %>% unique()

novel_ld_block_based_counts_all_gwas <- data.frame()
for (gwasid in gwas_ids) {
  novel_ld_block_based_counts_gwas <- novel_ld_block_based_counts_byqtl %>% filter(gwas_id == gwasid)
  qtlSubsets_missing_novel_lds <- qtlsubset_classes %>% 
    filter(!qtl_subset %in% novel_ld_block_based_counts_gwas$qtl_subset) %>% 
    mutate(gwas_id = unique(novel_ld_block_based_counts_gwas$gwas_id)) %>% 
    mutate(gwas_class = unique(novel_ld_block_based_counts_gwas$gwas_class)) %>% 
    mutate(sample_size_norm_novel_LD_block_count = 0)
  
  novel_ld_block_counts_all <- novel_ld_block_based_counts_gwas %>% rbind(qtlSubsets_missing_novel_lds)
  novel_ld_block_based_counts_all_gwas <- novel_ld_block_based_counts_all_gwas %>% rbind(novel_ld_block_counts_all)
}

novel_ld_block_based_counts_all_gwas <- novel_ld_block_based_counts_all_gwas %>% 
  mutate(gwas_id_short = gsub("1","", gsub("\\-.*","", gwas_id))) 

qtlsubsets_no_filter_to_label <- novel_ld_block_based_counts_all_gwas %>% 
  filter((gwas_id == "LC-ebi-a-GCST004627" & qtl_subset == "BLUEPRINT_PE.T-cell_ge") |
           (gwas_id == "MC-ebi-a-GCST004625" & qtl_subset == "Quach_2016.monocyte_naive_ge") |
           (gwas_id == "MPV-ebi-a-GCST004599" & qtl_subset =="CEDAR.platelet") |
           (gwas_id == "PLT-ebi-a-GCST004603" & qtl_subset == "CEDAR.platelet") |
           (gwas_id == "BMI-ukb-a-248" & qtl_subset == "HipSci.iPSC_ge") |
           (gwas_id == "RA-ebi-a-GCST002318" & qtl_subset == "BLUEPRINT_SE.monocyte_ge") |
           (gwas_id == "RA-ebi-a-GCST002318" & qtl_subset == "BLUEPRINT_PE.T-cell_ge") |
           (gwas_id == "IBD-ebi-a-GCST004131" & qtl_subset == "Schmiedel_2018.monocyte_naive_ge"))

write_tsv(novel_ld_block_based_counts_all_gwas, "../data/TSVs/colocs_merged_R3/novel_ld_blocks_classified.tsv")

temp_plot_violin_not_filter <- ggplot(novel_ld_block_based_counts_all_gwas, 
                           aes(x = sample_size_norm_novel_LD_block_count, 
                               y = gwas_id_short, 
                               label = qtl_subset)) +
    geom_violin(mapping = aes(color = present_in_GTEx,fill = present_in_GTEx), position="dodge", alpha=0.8) + 
    theme_bw() + 
    scale_color_manual(values = c("cadetblue3", "darkseagreen2"), 
                       name = "Cell type\npresence in GTEx", 
                       labels = c("Absent from GTEx", "Present in GTEx"))+
    scale_fill_manual(values = c("cadetblue3", "darkseagreen2"), 
                      name = "Cell type\npresence in GTEx", 
                      labels = c("Absent from GTEx", "Present in GTEx"))+
    # scale_color_brewer(palette = 'Set2') +
    # scale_fill_brewer(palette = 'Set2') +
    ggnewscale::new_scale_color() +
    geom_jitter(mapping = aes(color=cell_type_class, group = present_in_GTEx), 
                position=position_dodge(0.8), 
                size = 1, 
                shape = "triangle", 
                alpha=0.9) + 
    scale_color_manual(values = c("purple", "indianred1"), name = "Cell type class", labels = c("Immune cell types", "Others"))+
    facet_grid(gwas_class ~ ., shrink = TRUE, scales = "free_y", space = "free_y") +
    ggplot2::labs(x="Novel LD block count / Sample size", 
                  y="GWAS")
  
save_ggplots(plot = temp_plot_violin_not_filter, 
             path = paste0(output_dir, "/faceted_difference_violins/"), 
             filename = "GTExV8_vs_eQTLCatR3_violin_not_filter_no_label",width = 20)

temp_plot_violin_notfilt_labeled <- temp_plot_violin_not_filter +
  geom_label_repel(
      data = qtlsubsets_no_filter_to_label,
      aes(label = qtl_subset),
      size = 2,
      color = "black"
    ) 
save_ggplots(plot = temp_plot_violin_notfilt_labeled, 
             path = paste0(output_dir, "/faceted_difference_violins/"), 
             filename = "GTExV8_vs_eQTLCatR3_violin_not_filter_labeled",width = 20)

# ggplotly_plot <- plotly::ggplotly(temp_plot_violin)
# htmlwidgets::saveWidget(widget = plotly::as_widget(ggplotly_plot),
#                         file = file.path(normalizePath(paste0(output_dir, "/plotly")), "violin_plots.html"),
#                         libdir = "dependencies")
```


```{r}
novel_ld_block_based_counts_all_gwas_filt <- novel_ld_block_based_counts_all_gwas %>% filter(sample_size>150)

qtlsubsets_to_label <- novel_ld_block_based_counts_all_gwas_filt %>% 
    filter((gwas_id == "LC-ebi-a-GCST004627" & qtl_subset == "BLUEPRINT_PE.T-cell_ge") |
           (gwas_id == "MC-ebi-a-GCST004625" & qtl_subset == "Quach_2016.monocyte_naive_ge") |
           (gwas_id == "MPV-ebi-a-GCST004599" & qtl_subset =="CEDAR.platelet") |
           (gwas_id == "PLT-ebi-a-GCST004603" & qtl_subset == "CEDAR.platelet") |
           (gwas_id == "BMI-ukb-a-248" & qtl_subset == "HipSci.iPSC_ge") |
           (gwas_id == "RA-ebi-a-GCST002318" & qtl_subset == "BLUEPRINT_SE.monocyte_ge") |
           (gwas_id == "RA-ebi-a-GCST002318" & qtl_subset == "BLUEPRINT_PE.T-cell_ge"))

temp_plot_violin <- ggplot(novel_ld_block_based_counts_all_gwas_filt, 
                           aes(x = sample_size_norm_novel_LD_block_count, 
                               y = gwas_id_short, 
                               label = qtl_subset)) +
    geom_violin(mapping = aes(color = present_in_GTEx,fill = present_in_GTEx), position="dodge", alpha=0.8) + 
    theme_bw() + 
    scale_color_manual(values = c("cadetblue3", "darkseagreen2"), 
                       name = "Cell type\npresence in GTEx", 
                       labels = c("Absent from GTEx", "Present in GTEx"))+
    scale_fill_manual(values = c("cadetblue3", "darkseagreen2"), 
                      name = "Cell type\npresence in GTEx", 
                      labels = c("Absent from GTEx", "Present in GTEx"))+
    # scale_color_brewer(palette = 'Set2') +
    # scale_fill_brewer(palette = 'Set2') +
    ggnewscale::new_scale_color() +
    geom_jitter(mapping = aes(color=cell_type_class, group = present_in_GTEx), 
                position=position_dodge(0.8), 
                size = 1, 
                shape = "triangle", 
                alpha=0.9) + 
    scale_color_manual(values = c("purple", "indianred1"), name = "Cell type class", labels = c("Immune cell types", "Others"))+
    facet_grid(gwas_class ~ ., shrink = TRUE, scales = "free_y", space = "free_y") +
    ggplot2::labs(x="Novel LD block count / Sample size", 
                  y="GWAS")

save_ggplots(plot = temp_plot_violin, 
             path = paste0(output_dir, "/faceted_difference_violins/"), 
             filename = "GTExV8_vs_eQTLCatR3_violin_ss_filter_label_no_label",width = 20)

temp_plot_violin_labeled <- temp_plot_violin +
  geom_label_repel(
      data = qtlsubsets_to_label,
      aes(label = qtl_subset),
      size = 2,
      color = "black"
    ) 

save_ggplots(plot = temp_plot_violin_labeled, 
             path = paste0(output_dir, "/faceted_difference_violins/"), 
             filename = "GTExV8_vs_eQTLCatR3_violin_ss_filter_label",width = 20)
```











