---
title: "R Notebook"
output: html_notebook
---

```{r}
library("tidyverse")
library("ggplot2")
library("RColorBrewer")
library("ggrepel")
library("data.table")
library("plotly")
library("dplyr")
```

```{r}
ontology_map = readr::read_tsv("../../../ontology_mappings/tissue_ontology_mapping.tsv")
friendly_names = readr::read_tsv("../../../ontology_mappings/friendly_names.tsv") %>%
  dplyr::select(tissue_ontology_id, tissue_label)
ontology_map$study[c(1,2,3)] = c("BLUEPRINT_SE", "BLUEPRINT_SE", "BLUEPRINT_PE")
ontology_map <- ontology_map %>% dplyr::mutate(study_qtlgroup = paste0(study, ".", qtl_group)) %>%
  dplyr::left_join(friendly_names) %>% 
  dplyr::filter(tissue_ontology_id %in% c("EFO_0005292", "UBERON_0001013", "UBERON_0002097", 
                                          "UBERON_0000178", "UBERON_0001134", "CL_0000057",
                                          "UBERON_0009834", "CL_0000624", "CL_0002057")) %>% 
  dplyr::filter(study %in% c("BLUEPRINT_SE", "BLUEPRINT_PE","GENCORD", "GEUVADIS", "TwinsUK", 
                             "ROSMAP", "BrainSeq", "Lepik_2017", "FUSION","GTEx" ))

# LCLs - GTEx, TwinsUK, GENCORD, GEUVADIS
# Fat - FUSION, GTEx, TwinsUK
# Skin - TwinsUK, GTEx
# Blood - Lepik, GTEx, TwinsUK
# Muscle - GTEx, Fusion
# Fibroblast - GTEx, GENCORD
# Brain cortex - GTEx, RosMAP, BrainSeq
# T-cells 
# Monocyte

# friendly label
ontology_map = ontology_map %>% dplyr::mutate(label = paste(study, tissue_label, sep=" "))
```

```{r}
# txrev all together no graph
df_sharing = read_tsv("sharing_tsv/sharing_txrev_without_cc.tsv", 
                      col_names = c("dataset1","dataset2","value"), skip = 1)

df_sharing = df_sharing %>% 
    mutate(dataset1 = gsub(pattern = "GTExV8", replacement = "GTEx", x = dataset1)) %>% 
    mutate(dataset2 = gsub(pattern = "GTExV8", replacement = "GTEx", x = dataset2))

# remove diagonal values
df_sharing = df_sharing %>% filter(value < 1)
df_sharing_filt <- df_sharing %>% 
  filter(dataset1 %in% ontology_map$study_qtlgroup) %>% 
  filter(dataset2 %in% ontology_map$study_qtlgroup) %>% 
  group_by(value) %>% 
  dplyr::slice(1) %>% 
  ungroup()

df_sharing_filt <- df_sharing_filt %>% 
  left_join(ontology_map %>% select(study, study_qtlgroup, tissue_ontology_id, cell_type), by = c("dataset1" = "study_qtlgroup")) %>% 
  dplyr::rename(study1 = study) %>% 
  dplyr::rename(tissue_ontology_id1 = tissue_ontology_id) %>% 
  dplyr::rename(cell_type1 = cell_type) %>% 
  left_join(ontology_map %>% select(study, study_qtlgroup, tissue_ontology_id,cell_type), by = c("dataset2" = "study_qtlgroup")) %>% 
  dplyr::rename(study2 = study) %>% 
  dplyr::rename(tissue_ontology_id2 = tissue_ontology_id) %>% 
  dplyr::rename(cell_type2 = cell_type) 

df_sharing_filt <- df_sharing_filt %>% 
  dplyr::mutate(study1 = ifelse(study1 %in% c("BLUEPRINT_SE", "BLUEPRINT_PE"), "BLUEPRINT", study1)) %>% 
  dplyr::mutate(study2 = ifelse(study2 %in% c("BLUEPRINT_SE", "BLUEPRINT_PE"), "BLUEPRINT", study2)) 

df_sharing_filt <- df_sharing_filt %>% 
  mutate(same_study = study1==study2, same_celltype = tissue_ontology_id1==tissue_ontology_id2) 

df_sharing_filt <- df_sharing_filt %>% 
  filter(same_celltype | same_study)

df_sharing_filt <- df_sharing_filt %>% 
  mutate(label = ifelse(same_study & !same_celltype, "Same study,\ndifferent cell types", "Different studies,\nsame cell type"))

p <- ggplot(data = df_sharing_filt, mapping = aes(x=label, y=value, label = study1)) +
  geom_violin(trim=FALSE) +
  geom_dotplot(binaxis='y', stackdir='center', dotsize=1) +
  scale_y_continuous(breaks = seq(0, 1, 0.2)) +
  xlab("") +
  ylab("Mash sharing coefficient") +
  # geom_text_repel() +
  theme_bw() +   theme(axis.text=element_text(size=14)) 

ggsave(plot = p, 
       filename = "sharing_violin_plots/eqtl_sharing_celltype_vs_study_violin_txrev_no_cc.pdf", 
       device = "pdf", width = 5, height = 4)

# txrev all together no cc Mann-Whitney U test for the violin plots
wilcox_test <- wilcox.test(df_sharing_filt$value~df_sharing_filt$same_study)
annotate_label <- paste0("Two-sample Wilcoxon test p-value: ", wilcox_test$p.value)

p <- ggplot(data = df_sharing_filt, mapping = aes(x=label, y=value, label = study1)) +
  geom_violin(trim=FALSE) +
  geom_dotplot(binaxis='y', stackdir='center', dotsize=1) +
  # annotate(geom="label", x=-Inf, y=-Inf, hjust = -0.01, vjust=-0.5,label=annotate_label, size = 4, fill = "#ede8a4") +
  scale_y_continuous(breaks = seq(0, 1, 0.2)) +
  xlab("") +
  ylab("Mash sharing coefficient") +
  labs(title = "TxRevise all event types together, without graph", subtitle = annotate_label) +
  # geom_text_repel() +
  theme_bw() +   theme(axis.text=element_text(size=14)) 

ggsave(plot = p, 
       filename = "sharing_violin_plots/eqtl_sharing_celltype_vs_study_violin_txrev_no_cc_pval.pdf", 
       device = "pdf", width = 5, height = 4)

# df_sharing_filt <- df_sharing_filt %>%
#   mutate(summary_data = paste0(study1,"_",cell_type1,"_",study2,"_",cell_type2))
# 
# plot_for_plotly <- ggplot(data = df_sharing_filt, mapping = aes(x=label, y=value, label = summary_data)) + 
#   geom_point(mapping = aes(color = study1)) + 
#   theme_bw() +   theme(axis.text=element_text(size=14)) 
# ggplotly(plot_for_plotly)

wilcox.test(df_sharing_filt$value~df_sharing_filt$same_study)
# txrev all together Mann-Whitney U test for the violin plots
```

```{r}
# gene expression connected components thresh 30
df_sharing = read_tsv("sharing_tsv/sharing_ge_with_cc_30.tsv", 
                      col_names = c("dataset1","dataset2","value"), skip = 1)

df_sharing = df_sharing %>% 
    mutate(dataset1 = gsub(pattern = "GTExV8", replacement = "GTEx", x = dataset1)) %>% 
    mutate(dataset2 = gsub(pattern = "GTExV8", replacement = "GTEx", x = dataset2))

# remove diagonal values
df_sharing = df_sharing %>% filter(value < 1)
df_sharing_filt <- df_sharing %>% 
  filter(dataset1 %in% ontology_map$study_qtlgroup) %>% 
  filter(dataset2 %in% ontology_map$study_qtlgroup) %>% 
  group_by(value) %>% 
  dplyr::slice(1) %>% 
  ungroup()

df_sharing_filt <- df_sharing_filt %>% 
  left_join(ontology_map %>% select(study, study_qtlgroup, tissue_ontology_id, cell_type), by = c("dataset1" = "study_qtlgroup")) %>% 
  dplyr::rename(study1 = study) %>% 
  dplyr::rename(tissue_ontology_id1 = tissue_ontology_id) %>% 
  dplyr::rename(cell_type1 = cell_type) %>% 
  left_join(ontology_map %>% select(study, study_qtlgroup, tissue_ontology_id,cell_type), by = c("dataset2" = "study_qtlgroup")) %>% 
  dplyr::rename(study2 = study) %>% 
  dplyr::rename(tissue_ontology_id2 = tissue_ontology_id) %>% 
  dplyr::rename(cell_type2 = cell_type) 

df_sharing_filt <- df_sharing_filt %>% 
  dplyr::mutate(study1 = ifelse(study1 %in% c("BLUEPRINT_SE", "BLUEPRINT_PE"), "BLUEPRINT", study1)) %>% 
  dplyr::mutate(study2 = ifelse(study2 %in% c("BLUEPRINT_SE", "BLUEPRINT_PE"), "BLUEPRINT", study2)) 

df_sharing_filt <- df_sharing_filt %>% 
  mutate(same_study = study1==study2, same_celltype = tissue_ontology_id1==tissue_ontology_id2) 

df_sharing_filt <- df_sharing_filt %>% 
  filter(same_celltype | same_study)

df_sharing_filt <- df_sharing_filt %>% 
  mutate(label = ifelse(same_study & !same_celltype, "Same study,\ndifferent cell types", "Different studies,\nsame cell type"))


p <- ggplot(data = df_sharing_filt, mapping = aes(x=label, y=value, label = study1)) +
  geom_violin(trim=FALSE) +
  geom_dotplot(binaxis='y', stackdir='center', dotsize=1) +
  scale_y_continuous(breaks = seq(0, 1, 0.2)) +
  xlab("") +
  ylab("Mash sharing coefficient") +
  # geom_text_repel() +
  theme_bw() +   theme(axis.text=element_text(size=14)) 

ggsave(plot = p, 
       filename = "sharing_violin_plots/eqtl_sharing_celltype_vs_study_violin_ge_thresh_30_with_cc.pdf", 
       device = "pdf", width = 5, height = 4)

# Gene expression (connected components approach) Mann-Whitney U test for the violin plots
wilcox_test <- wilcox.test(df_sharing_filt$value~df_sharing_filt$same_study)
annotate_label <- paste0("Two-sample Wilcoxon test p-value: ", wilcox_test$p.value)

p <- ggplot(data = df_sharing_filt, mapping = aes(x=label, y=value, label = study1)) +
  geom_violin(trim=FALSE) +
  geom_dotplot(binaxis='y', stackdir='center', dotsize=1) +
  # annotate(geom="label", x=-Inf, y=-Inf, hjust = -0.01, vjust=-0.5,label=annotate_label, size = 4, fill = "#ede8a4") +
  scale_y_continuous(breaks = seq(0, 1, 0.2)) +
  xlab("") +
  ylab("Mash sharing coefficient") +
  labs(title = "Gene expression with graph (cs_size threshold = 30)", subtitle = annotate_label) +
  # geom_text_repel() +
  theme_bw() +   theme(axis.text=element_text(size=14)) 

ggsave(plot = p, 
       filename = "sharing_violin_plots//eqtl_sharing_celltype_vs_study_violin_ge_thresh_30_with_cc_pval.pdf", 
       device = "pdf", width = 5, height = 4)

# df_sharing_filt <- df_sharing_filt %>%
#   mutate(summary_data = paste0(study1,"_",cell_type1,"_",study2,"_",cell_type2))
# 
# plot_for_plotly <- ggplot(data = df_sharing_filt, mapping = aes(x=label, y=value, label = summary_data)) + 
#   geom_point(mapping = aes(color = study1)) + 
#   theme_bw() +   theme(axis.text=element_text(size=14)) 
# ggplotly_plot <- ggplotly(plot_for_plotly)
# 
# htmlwidgets::saveWidget(widget = plotly::as_widget(ggplotly_plot),
#                         file = paste0(path_file,"/eqtl_sharing_celltype_vs_study_violin.html"),
#                         libdir = "dependencies")
wilcox_test
```

```{r}
# gene expression connected components thresh 30 with min p-val selection
df_sharing = read_tsv("sharing_tsv/sharing_ge_with_cc_30_minpval.tsv", 
                      col_names = c("dataset1","dataset2","value"), skip = 1)

df_sharing = df_sharing %>% 
    mutate(dataset1 = gsub(pattern = "GTExV8", replacement = "GTEx", x = dataset1)) %>% 
    mutate(dataset2 = gsub(pattern = "GTExV8", replacement = "GTEx", x = dataset2))

# remove diagonal values
df_sharing = df_sharing %>% filter(value < 1)
df_sharing_filt <- df_sharing %>% 
  filter(dataset1 %in% ontology_map$study_qtlgroup) %>% 
  filter(dataset2 %in% ontology_map$study_qtlgroup) %>% 
  group_by(value) %>% 
  dplyr::slice(1) %>% 
  ungroup()

df_sharing_filt <- df_sharing_filt %>% 
  left_join(ontology_map %>% select(study, study_qtlgroup, tissue_ontology_id, cell_type), by = c("dataset1" = "study_qtlgroup")) %>% 
  dplyr::rename(study1 = study) %>% 
  dplyr::rename(tissue_ontology_id1 = tissue_ontology_id) %>% 
  dplyr::rename(cell_type1 = cell_type) %>% 
  left_join(ontology_map %>% select(study, study_qtlgroup, tissue_ontology_id,cell_type), by = c("dataset2" = "study_qtlgroup")) %>% 
  dplyr::rename(study2 = study) %>% 
  dplyr::rename(tissue_ontology_id2 = tissue_ontology_id) %>% 
  dplyr::rename(cell_type2 = cell_type) 

df_sharing_filt <- df_sharing_filt %>% 
  dplyr::mutate(study1 = ifelse(study1 %in% c("BLUEPRINT_SE", "BLUEPRINT_PE"), "BLUEPRINT", study1)) %>% 
  dplyr::mutate(study2 = ifelse(study2 %in% c("BLUEPRINT_SE", "BLUEPRINT_PE"), "BLUEPRINT", study2)) 

df_sharing_filt <- df_sharing_filt %>% 
  mutate(same_study = study1==study2, same_celltype = tissue_ontology_id1==tissue_ontology_id2) 

df_sharing_filt <- df_sharing_filt %>% 
  filter(same_celltype | same_study)

df_sharing_filt <- df_sharing_filt %>% 
  mutate(label = ifelse(same_study & !same_celltype, "Same study,\ndifferent cell types", "Different studies,\nsame cell type"))


p <- ggplot(data = df_sharing_filt, mapping = aes(x=label, y=value, label = study1)) +
  geom_violin(trim=FALSE) +
  geom_dotplot(binaxis='y', stackdir='center', dotsize=1) +
  scale_y_continuous(breaks = seq(0, 1, 0.2)) +
  xlab("") +
  ylab("Mash sharing coefficient") +
  # geom_text_repel() +
  theme_bw() +   theme(axis.text=element_text(size=14)) 

ggsave(plot = p, 
       filename = "sharing_violin_plots/eqtl_sharing_celltype_vs_study_violin_ge_thresh_30_with_cc_minpval.pdf", 
       device = "pdf", width = 5, height = 4)

# Gene expression (connected components approach) Mann-Whitney U test for the violin plots
wilcox_test <- wilcox.test(df_sharing_filt$value~df_sharing_filt$same_study)
annotate_label <- paste0("Two-sample Wilcoxon test p-value: ", wilcox_test$p.value)

p <- ggplot(data = df_sharing_filt, mapping = aes(x=label, y=value, label = study1)) +
  geom_violin(trim=FALSE) +
  geom_dotplot(binaxis='y', stackdir='center', dotsize=1) +
  # annotate(geom="label", x=-Inf, y=-Inf, hjust = -0.01, vjust=-0.5,label=annotate_label, size = 4, fill = "#ede8a4") +
  scale_y_continuous(breaks = seq(0, 1, 0.2)) +
  xlab("") +
  ylab("Mash sharing coefficient") +
  labs(title = "GE with graph and minpval (cs_size threshold = 30)", subtitle = annotate_label) +
  # geom_text_repel() +
  theme_bw() +   theme(axis.text=element_text(size=14)) 

ggsave(plot = p, 
       filename = "sharing_violin_plots//eqtl_sharing_celltype_vs_study_violin_ge_thresh_30_with_cc_pval_minpval.pdf", 
       device = "pdf", width = 5, height = 4)

# df_sharing_filt <- df_sharing_filt %>%
#   mutate(summary_data = paste0(study1,"_",cell_type1,"_",study2,"_",cell_type2))
# 
# plot_for_plotly <- ggplot(data = df_sharing_filt, mapping = aes(x=label, y=value, label = summary_data)) + 
#   geom_point(mapping = aes(color = study1)) + 
#   theme_bw() +   theme(axis.text=element_text(size=14)) 
# ggplotly_plot <- ggplotly(plot_for_plotly)
# 
# htmlwidgets::saveWidget(widget = plotly::as_widget(ggplotly_plot),
#                         file = paste0(path_file,"/eqtl_sharing_celltype_vs_study_violin.html"),
#                         libdir = "dependencies")
wilcox_test
```


```{r}
# gene expression connected components thresh 15
df_sharing = read_tsv("sharing_tsv/sharing_ge_with_cc_15.tsv", 
                      col_names = c("dataset1","dataset2","value"), skip = 1)

df_sharing = df_sharing %>% 
    mutate(dataset1 = gsub(pattern = "GTExV8", replacement = "GTEx", x = dataset1)) %>% 
    mutate(dataset2 = gsub(pattern = "GTExV8", replacement = "GTEx", x = dataset2))

# remove diagonal values
df_sharing = df_sharing %>% filter(value < 1)
df_sharing_filt <- df_sharing %>% 
  filter(dataset1 %in% ontology_map$study_qtlgroup) %>% 
  filter(dataset2 %in% ontology_map$study_qtlgroup) %>% 
  group_by(value) %>% 
  dplyr::slice(1) %>% 
  ungroup()

df_sharing_filt <- df_sharing_filt %>% 
  left_join(ontology_map %>% select(study, study_qtlgroup, tissue_ontology_id, cell_type), by = c("dataset1" = "study_qtlgroup")) %>% 
  dplyr::rename(study1 = study) %>% 
  dplyr::rename(tissue_ontology_id1 = tissue_ontology_id) %>% 
  dplyr::rename(cell_type1 = cell_type) %>% 
  left_join(ontology_map %>% select(study, study_qtlgroup, tissue_ontology_id,cell_type), by = c("dataset2" = "study_qtlgroup")) %>% 
  dplyr::rename(study2 = study) %>% 
  dplyr::rename(tissue_ontology_id2 = tissue_ontology_id) %>% 
  dplyr::rename(cell_type2 = cell_type) 

df_sharing_filt <- df_sharing_filt %>% 
  dplyr::mutate(study1 = ifelse(study1 %in% c("BLUEPRINT_SE", "BLUEPRINT_PE"), "BLUEPRINT", study1)) %>% 
  dplyr::mutate(study2 = ifelse(study2 %in% c("BLUEPRINT_SE", "BLUEPRINT_PE"), "BLUEPRINT", study2)) 

df_sharing_filt <- df_sharing_filt %>% 
  mutate(same_study = study1==study2, same_celltype = tissue_ontology_id1==tissue_ontology_id2) 

df_sharing_filt <- df_sharing_filt %>% 
  filter(same_celltype | same_study)

df_sharing_filt <- df_sharing_filt %>% 
  mutate(label = ifelse(same_study & !same_celltype, "Same study,\ndifferent cell types", "Different studies,\nsame cell type"))


p <- ggplot(data = df_sharing_filt, mapping = aes(x=label, y=value, label = study1)) +
  geom_violin(trim=FALSE) +
  geom_dotplot(binaxis='y', stackdir='center', dotsize=1) +
  scale_y_continuous(breaks = seq(0, 1, 0.2)) +
  xlab("") +
  ylab("Mash sharing coefficient") +
  # geom_text_repel() +
  theme_bw() +   theme(axis.text=element_text(size=14)) 

ggsave(plot = p, 
       filename = "sharing_violin_plots/eqtl_sharing_celltype_vs_study_violin_ge_thresh_15_with_cc.pdf", 
       device = "pdf", width = 5, height = 4)

# Gene expression (connected components approach) Mann-Whitney U test for the violin plots
wilcox_test <- wilcox.test(df_sharing_filt$value~df_sharing_filt$same_study)
annotate_label <- paste0("Two-sample Wilcoxon test p-value: ", wilcox_test$p.value)

p <- ggplot(data = df_sharing_filt, mapping = aes(x=label, y=value, label = study1)) +
  geom_violin(trim=FALSE) +
  geom_dotplot(binaxis='y', stackdir='center', dotsize=1) +
  # annotate(geom="label", x=-Inf, y=-Inf, hjust = -0.01, vjust=-0.5,label=annotate_label, size = 4, fill = "#ede8a4") +
  scale_y_continuous(breaks = seq(0, 1, 0.2)) +
  xlab("") +
  ylab("Mash sharing coefficient") +
  labs(title = "Gene expression with graph (cs_size threshold = 15)", subtitle = annotate_label) +
  # geom_text_repel() +
  theme_bw() +   theme(axis.text=element_text(size=14)) 

ggsave(plot = p, 
       filename = "sharing_violin_plots//eqtl_sharing_celltype_vs_study_violin_ge_thresh_15_with_cc_pval.pdf", 
       device = "pdf", width = 5, height = 4)

# df_sharing_filt <- df_sharing_filt %>%
#   mutate(summary_data = paste0(study1,"_",cell_type1,"_",study2,"_",cell_type2))
# 
# plot_for_plotly <- ggplot(data = df_sharing_filt, mapping = aes(x=label, y=value, label = summary_data)) + 
#   geom_point(mapping = aes(color = study1)) + 
#   theme_bw() +   theme(axis.text=element_text(size=14)) 
# ggplotly_plot <- ggplotly(plot_for_plotly)
# 
# htmlwidgets::saveWidget(widget = plotly::as_widget(ggplotly_plot),
#                         file = paste0(path_file,"/eqtl_sharing_celltype_vs_study_violin.html"),
#                         libdir = "dependencies")
wilcox_test
```


Microarray studies
```{r}
# ontology_map = readr::read_tsv("../../GitHub/eQTL-Catalogue-resources/ontology_mappings/tissue_ontology_mapping.tsv")
# friendly_names = readr::read_tsv("../../GitHub/eQTL-Catalogue-resources/ontology_mappings/friendly_names.tsv") %>%
#   dplyr::select(tissue_ontology_id, tissue_label)
# ontology_map$study[c(1,2,3)] = c("BLUEPRINT_SE", "BLUEPRINT_SE", "BLUEPRINT_PE")
# ontology_map_microarray <- ontology_map %>% dplyr::mutate(study_qtlgroup = paste0(study, ".", qtl_group)) %>%
#   dplyr::left_join(friendly_names) %>% 
#   dplyr::filter(study %in% c("Fairfax_2014", "CEDAR","Fairfax_2012", "Naranbhai_2015", "Kasela_2017" ))

# Monocytes (CL_0002057), CD4T (CL_0000624), CD8T (CL_0000625), neutrophils (CL_0000775), B-cells (CL_0000236)

# ontology_map_microarray <- ontology_map_microarray %>% 
#   dplyr::filter(tissue_ontology_id %in% c("CL_0002057", "CL_0000624", "CL_0000625", 
#                                           "CL_0000775", "CL_0000236"))

# friendly label
# ontology_map = ontology_map %>% dplyr::mutate(label = paste(study, tissue_label, sep=" "))
```

```{r}
# transcript usage SHARING

df_sharing = read_tsv("sharing_tsv/sharing_tx_without_cc.tsv", 
                      col_names = c("dataset1","dataset2","value"), skip = 1)

df_sharing = df_sharing %>% 
    mutate(dataset1 = gsub(pattern = "GTExV8", replacement = "GTEx", x = dataset1)) %>% 
    mutate(dataset2 = gsub(pattern = "GTExV8", replacement = "GTEx", x = dataset2))

# remove diagonal values
df_sharing = df_sharing %>% filter(value < 1)
df_sharing_filt <- df_sharing %>% 
  filter(dataset1 %in% ontology_map$study_qtlgroup) %>% 
  filter(dataset2 %in% ontology_map$study_qtlgroup) %>% 
  group_by(value) %>% 
  dplyr::slice(1) %>% 
  ungroup()

df_sharing_filt <- df_sharing_filt %>% 
  left_join(ontology_map %>% select(study, study_qtlgroup, tissue_ontology_id, cell_type), by = c("dataset1" = "study_qtlgroup")) %>% 
  dplyr::rename(study1 = study) %>% 
  dplyr::rename(tissue_ontology_id1 = tissue_ontology_id) %>% 
  dplyr::rename(cell_type1 = cell_type) %>% 
  left_join(ontology_map %>% select(study, study_qtlgroup, tissue_ontology_id,cell_type), by = c("dataset2" = "study_qtlgroup")) %>% 
  dplyr::rename(study2 = study) %>% 
  dplyr::rename(tissue_ontology_id2 = tissue_ontology_id) %>% 
  dplyr::rename(cell_type2 = cell_type) 

df_sharing_filt <- df_sharing_filt %>% 
  dplyr::mutate(study1 = ifelse(study1 %in% c("BLUEPRINT_SE", "BLUEPRINT_PE"), "BLUEPRINT", study1)) %>% 
  dplyr::mutate(study2 = ifelse(study2 %in% c("BLUEPRINT_SE", "BLUEPRINT_PE"), "BLUEPRINT", study2)) 


df_sharing_filt <- df_sharing_filt %>% 
  dplyr::mutate(same_study = study1==study2, same_celltype = tissue_ontology_id1==tissue_ontology_id2) 

df_sharing_filt <- df_sharing_filt %>% 
  dplyr::filter(same_celltype | same_study)

df_sharing_filt <- df_sharing_filt %>% 
  dplyr::mutate(label = ifelse(same_study & !same_celltype, "Same study,\ndifferent cell types", "Different studies,\nsame cell type"))

p <- ggplot(data = df_sharing_filt, mapping = aes(x=label, y=value, label = study1)) +
  geom_violin(trim=FALSE) +
  geom_dotplot(binaxis='y', stackdir='center', dotsize=1) +
  scale_y_continuous(breaks = seq(0, 1, 0.2)) +
  xlab("") +
  ylab("Mash sharing coefficient") +
  # geom_text_repel() +
  theme_bw() +   theme(axis.text=element_text(size=14)) 

ggsave(plot = p, 
       filename = "sharing_violin_plots//eqtl_sharing_celltype_vs_study_violin_tx_no_cc.pdf", 
       device = "pdf", width = 5, height = 4)

# tx usage no cc Mann-Whitney U test for the violin plots
wilcox_test <- wilcox.test(df_sharing_filt$value~df_sharing_filt$same_study)
annotate_label <- paste0("Two-sample Wilcoxon test p-value: ", wilcox_test$p.value)

p <- ggplot(data = df_sharing_filt, mapping = aes(x=label, y=value, label = study1)) +
  geom_violin(trim=FALSE) +
  geom_dotplot(binaxis='y', stackdir='center', dotsize=1) +
  # annotate(geom="label", x=-Inf, y=-Inf, hjust = -0.01, vjust=-0.5,label=annotate_label, size = 4, fill = "#ede8a4") +
  scale_y_continuous(breaks = seq(0, 1, 0.2)) +
  xlab("") +
  ylab("Mash sharing coefficient") +
  labs(title = "Transcript usage without graph", subtitle = annotate_label) +
  # geom_text_repel() +
  theme_bw() +   theme(axis.text=element_text(size=14)) 

ggsave(plot = p, 
       filename = "sharing_violin_plots//eqtl_sharing_celltype_vs_study_violin_tx_no_cc_pval.pdf", 
       device = "pdf", width = 5, height = 4)

# df_sharing_filt <- df_sharing_filt %>%
#   mutate(summary_data = paste0(study1,"_",cell_type1,"_",study2,"_",cell_type2))
# 
# plot_for_plotly <- ggplot(data = df_sharing_filt, mapping = aes(x=label, y=value, label = summary_data)) + 
#   geom_point(mapping = aes(color = study1)) + 
#   theme_bw() +   theme(axis.text=element_text(size=14)) 
# ggplotly(plot_for_plotly)
# 
# write_tsv(df_sharing_filt, "../data/eqtl_sharing_study_celltype.tsv")

# Transcript usage Mann-Whitney U test for the violin plots
wilcox.test(df_sharing_filt$value~df_sharing_filt$same_study)
```

```{r}
# EXON expression
df_sharing = read_tsv("sharing_tsv/sharing_exon_without_cc.tsv", 
                col_names = c("dataset1","dataset2","value"), skip = 1)

df_sharing = df_sharing %>% 
    mutate(dataset1 = gsub(pattern = "GTExV8", replacement = "GTEx", x = dataset1)) %>% 
    mutate(dataset2 = gsub(pattern = "GTExV8", replacement = "GTEx", x = dataset2))


# remove diagonal values
df_sharing = df_sharing %>% filter(value < 1)
df_sharing_filt <- df_sharing %>% 
  filter(dataset1 %in% ontology_map$study_qtlgroup) %>% 
  filter(dataset2 %in% ontology_map$study_qtlgroup) %>% 
  group_by(value) %>% 
  dplyr::slice(1) %>% 
  ungroup()

df_sharing_filt <- df_sharing_filt %>% 
  left_join(ontology_map %>% select(study, study_qtlgroup, tissue_ontology_id, cell_type), by = c("dataset1" = "study_qtlgroup")) %>% 
  dplyr::rename(study1 = study) %>% 
  dplyr::rename(tissue_ontology_id1 = tissue_ontology_id) %>% 
  dplyr::rename(cell_type1 = cell_type) %>% 
  left_join(ontology_map %>% select(study, study_qtlgroup, tissue_ontology_id,cell_type), by = c("dataset2" = "study_qtlgroup")) %>% 
  dplyr::rename(study2 = study) %>% 
  dplyr::rename(tissue_ontology_id2 = tissue_ontology_id) %>% 
  dplyr::rename(cell_type2 = cell_type) 

df_sharing_filt <- df_sharing_filt %>% 
  dplyr::mutate(study1 = ifelse(study1 %in% c("BLUEPRINT_SE", "BLUEPRINT_PE"), "BLUEPRINT", study1)) %>% 
  dplyr::mutate(study2 = ifelse(study2 %in% c("BLUEPRINT_SE", "BLUEPRINT_PE"), "BLUEPRINT", study2)) 

df_sharing_filt <- df_sharing_filt %>% 
  mutate(same_study = study1==study2, same_celltype = tissue_ontology_id1==tissue_ontology_id2) 

df_sharing_filt <- df_sharing_filt %>% 
  filter(same_celltype | same_study)

df_sharing_filt <- df_sharing_filt %>% 
  mutate(label = ifelse(same_study & !same_celltype, "Same study,\ndifferent cell types", "Different studies,\nsame cell type"))


p <- ggplot(data = df_sharing_filt, mapping = aes(x=label, y=value, label = study1)) +
  geom_violin(trim=FALSE) +
  geom_dotplot(binaxis='y', stackdir='center', dotsize=1) +
  scale_y_continuous(breaks = seq(0, 1, 0.2)) +
  xlab("") +
  ylab("Mash sharing coefficient") +
  # geom_text_repel() +
  theme_bw() +   theme(axis.text=element_text(size=14)) 

ggsave(plot = p, 
       filename = "sharing_violin_plots//eqtl_sharing_celltype_vs_study_violin_exon_no_cc.pdf", 
       device = "pdf", width = 5, height = 4)

# exon expression (connected components approach) Mann-Whitney U test for the violin plots
wilcox_test <- wilcox.test(df_sharing_filt$value~df_sharing_filt$same_study)
annotate_label <- paste0("Two-sample Wilcoxon test p-value: ", wilcox_test$p.value)

p <- ggplot(data = df_sharing_filt, mapping = aes(x=label, y=value, label = study1)) +
  geom_violin(trim=FALSE) +
  geom_dotplot(binaxis='y', stackdir='center', dotsize=1) +
  # annotate(geom="label", x=-Inf, y=-Inf, hjust = -0.01, vjust=-0.5,label=annotate_label, size = 4, fill = "#ede8a4") +
  scale_y_continuous(breaks = seq(0, 1, 0.2)) +
  xlab("") +
  ylab("Mash sharing coefficient") +
  labs(title = "Exon expression without graph", subtitle = annotate_label) +
  # geom_text_repel() +
  theme_bw() +   theme(axis.text=element_text(size=14)) 

ggsave(plot = p, 
       filename = "sharing_violin_plots//eqtl_sharing_celltype_vs_study_violin_exon_no_cc_pval.pdf", 
       device = "pdf", width = 5, height = 4)

# df_sharing_filt <- df_sharing_filt %>%
#   mutate(summary_data = paste0(study1,"_",cell_type1,"_",study2,"_",cell_type2))
# 
# plot_for_plotly <- ggplot(data = df_sharing_filt, mapping = aes(x=label, y=value, label = summary_data)) + 
#   geom_point(mapping = aes(color = study1)) + 
#   theme_bw() +   theme(axis.text=element_text(size=14)) 
# ggplotly(plot_for_plotly)
# 
# # write_tsv(df_sharing_filt, "../data/eqtl_sharing_study_celltype.tsv")
# 
# # Exon expression Mann-Whitney U test for the violin plots
wilcox.test(df_sharing_filt$value~df_sharing_filt$same_study)
```

```{r}
# GENE SHARING
df_sharing = read_tsv("sharing_tsv/sharing_ge_without_cc.tsv", 
                col_names = c("dataset1","dataset2","value"), skip = 1)

df_sharing = df_sharing %>% 
    mutate(dataset1 = gsub(pattern = "GTExV8", replacement = "GTEx", x = dataset1)) %>% 
    mutate(dataset2 = gsub(pattern = "GTExV8", replacement = "GTEx", x = dataset2))

# remove diagonal values
df_sharing = df_sharing %>% filter(value < 1)
df_sharing_filt <- df_sharing %>% 
  filter(dataset1 %in% ontology_map$study_qtlgroup) %>% 
  filter(dataset2 %in% ontology_map$study_qtlgroup) %>% 
  group_by(value) %>% 
  dplyr::slice(1) %>% 
  ungroup()

df_sharing_filt <- df_sharing_filt %>% 
  left_join(ontology_map %>% select(study, study_qtlgroup, tissue_ontology_id, cell_type), by = c("dataset1" = "study_qtlgroup")) %>% 
  dplyr::rename(study1 = study) %>% 
  dplyr::rename(tissue_ontology_id1 = tissue_ontology_id) %>% 
  dplyr::rename(cell_type1 = cell_type) %>% 
  left_join(ontology_map %>% select(study, study_qtlgroup, tissue_ontology_id,cell_type), by = c("dataset2" = "study_qtlgroup")) %>% 
  dplyr::rename(study2 = study) %>% 
  dplyr::rename(tissue_ontology_id2 = tissue_ontology_id) %>% 
  dplyr::rename(cell_type2 = cell_type) 

df_sharing_filt <- df_sharing_filt %>% 
  dplyr::mutate(study1 = ifelse(study1 %in% c("BLUEPRINT_SE", "BLUEPRINT_PE"), "BLUEPRINT", study1)) %>% 
  dplyr::mutate(study2 = ifelse(study2 %in% c("BLUEPRINT_SE", "BLUEPRINT_PE"), "BLUEPRINT", study2)) 

df_sharing_filt <- df_sharing_filt %>% 
  mutate(same_study = study1==study2, same_celltype = tissue_ontology_id1==tissue_ontology_id2) 

df_sharing_filt <- df_sharing_filt %>% 
  filter(same_celltype | same_study)

df_sharing_filt <- df_sharing_filt %>% 
  mutate(label = ifelse(same_study & !same_celltype, "Same study,\ndifferent cell types", "Different studies,\nsame cell type"))


p <- ggplot(data = df_sharing_filt, mapping = aes(x=label, y=value, label = study1)) +
  geom_violin(trim=FALSE) +
  geom_dotplot(binaxis='y', stackdir='center', dotsize=1) +
  scale_y_continuous(breaks = seq(0, 1, 0.2)) +
  xlab("") +
  ylab("Mash sharing coefficient") +
  # geom_text_repel() +
  theme_bw() +   theme(axis.text=element_text(size=14)) 

ggsave(plot = p, 
       filename = "sharing_violin_plots//eqtl_sharing_celltype_vs_study_violin_ge_no_cc.pdf", 
       device = "pdf", width = 5, height = 4)

# gene expression (connected components approach) Mann-Whitney U test for the violin plots
wilcox_test <- wilcox.test(df_sharing_filt$value~df_sharing_filt$same_study)
annotate_label <- paste0("Two-sample Wilcoxon test p-value: ", wilcox_test$p.value)

p <- ggplot(data = df_sharing_filt, mapping = aes(x=label, y=value, label = study1)) +
  geom_violin(trim=FALSE) +
  geom_dotplot(binaxis='y', stackdir='center', dotsize=1) +
  # annotate(geom="label", x=-Inf, y=-Inf, hjust = -0.01, vjust=-0.5,label=annotate_label, size = 4, fill = "#ede8a4") +
  scale_y_continuous(breaks = seq(0, 1, 0.2)) +
  xlab("") +
  ylab("Mash sharing coefficient") +
  labs(title = "Gene expression without graph", subtitle = annotate_label) +
  theme_bw() +   theme(axis.text=element_text(size=14)) 

ggsave(plot = p, 
       filename = "sharing_violin_plots//eqtl_sharing_celltype_vs_study_violin_ge_no_cc_pval.pdf", 
       device = "pdf", width = 5, height = 4)

# df_sharing_filt <- df_sharing_filt %>%
#   mutate(summary_data = paste0(study1,"_",cell_type1,"_",study2,"_",cell_type2))
# 
# plot_for_plotly <- ggplot(data = df_sharing_filt, mapping = aes(x=label, y=value, label = summary_data)) + 
#   geom_point(mapping = aes(color = study1)) + 
#   theme_bw() +   theme(axis.text=element_text(size=14)) 
# ggplotly(plot_for_plotly)

# write_tsv(df_sharing_filt, "../data/eqtl_sharing_study_celltype.tsv")
wilcox_test
```



