---
title: "R Notebook"
output: html_notebook
---

First things first: Loading needed libraries
```{r load_libraries, warning=FALSE, message=FALSE}
library("dplyr")
library("ggplot2")
library("readr")
library("stringr")
library("httr")
library("jsonlite")
library("tidyverse")
library("coloc")
library("biomaRt")
library("wiggleplotr")
library("GenomicRanges")
```

We defined a local function to easily fetch data from both GWAS and eQTL APIs
```{r define_function_to_fetch_paginated_API_response, warning=FALSE}
fetch_from_eqtl_cat_API <- function(link, is_gwas = FALSE){
  nullToNA <- function(x) {
    x[sapply(x, is.null)] <- NA
    return(x)
  }
  if (is_gwas) {
    cols_to_nest <- c("base_pair_location", "variant_id", "beta", "effect_allele_frequency", 
                      "ci_lower", "ci_upper", "other_allele", "odds_ratio", "p_value", 
                      "study_accession", "chromosome", "code", "effect_allele", "trait")
  } else {
    cols_to_nest <- c("alt", "position", "beta", "an", "ref", "variant", "pvalue", "chromosome", 
                      "rsid", "maf", "study_id", "median_tpm", "r2", "type", "qtl_group", "tissue_label", 
                      "condition", "ac", "condition_label", "molecular_trait_id", "gene_id", "tissue")
  }
  is_paginated <- !str_detect(link,"paginate=False")
  message("isPagined:", is_paginated)
  page = 1
  merged_summaries <- data.frame()
  while(!is.null(link)){
    print(paste0("Fetching page #",page))
    api_raw_data <- fromJSON(link, simplifyDataFrame = TRUE, flatten = TRUE)
    link <- api_raw_data$`_links`$`next`$href
    if (is_empty(api_raw_data$`_embedded`$associations)) {
      return(merged_summaries)
    }
    eqtl_raw_list_data <- do.call(rbind, lapply(api_raw_data$`_embedded`$associations, rbind))
    eqtl_data <- nullToNA(eqtl_raw_list_data) %>% as.matrix() %>% as_tibble()
    if (is_paginated) { eqtl_data <- dplyr::select(eqtl_data, -c("_links")) }
    eqtl_data <- tidyr::unnest(eqtl_data, cols = cols_to_nest)
    if (!is.null(link)) {
      page <- page + 1
    }
    merged_summaries <- merged_summaries %>% rbind(eqtl_data)
  }
  return(merged_summaries)
}
```

We first fetch GWAS data
========================
Chromosome: 20
study_accession: GCST002318
bp_lower: 45980000
bp_upper: 46200000
```{r fetch_GWAS_data}
RA_gwas_query_str <- "https://www.ebi.ac.uk/gwas/summary-statistics/api/chromosomes/20/associations?study_accession=GCST002318&bp_lower=45980000&bp_upper=46200000&size=1000"
gwas_data <- fetch_from_eqtl_cat_API(link = RA_gwas_query_str, is_gwas = TRUE)
message("Downloaded ", gwas_data %>% nrow(), " associations from GWAS Catalogue")
gwas_data <- gwas_data %>% 
  dplyr::filter(!ci_lower %>% is.na()) %>% 
  dplyr::filter(!ci_upper %>% is.na()) %>% 
  dplyr::mutate(log_OR = log(odds_ratio)) %>%
  dplyr::mutate(se = (log(ci_upper)-log(ci_lower))/3.92)
message("There remain ", gwas_data %>% nrow(), " associations after filtering invalid values")
```

Pick the lead variant from GWAS data
```{r}
lead_var_gwas = dplyr::arrange(gwas_data, p_value)[1,]
```
Pick significant associations in eQTL Catalogue with lead variant ID
```{r}
# "ROSMAP"
# "Schmiedel_2018",
rnaseq_study_names <- c("Alasoo_2018", "BLUEPRINT", "BrainSeq", "GENCORD", "GEUVADIS", "HipSci", "Lepik_2017", "Nedelec_2016", "Quach_2016", "Schwartzentruber_2018", "TwinsUK", "van_de_Bunt_2015")
rnaseq_studies_sign_assocs <- data.frame()
for(study_name in rnaseq_study_names){
  print(study_name)
  ge_study_query_gwas_lead <- paste0("https://www.ebi.ac.uk/eqtl/api/associations?variant_id=", lead_var_gwas$variant_id,
                                     "&p_upper=0.0001&study=",study_name)
  fetched_df <- fetch_from_eqtl_cat_API(ge_study_query_gwas_lead)
  rnaseq_studies_sign_assocs <- rnaseq_studies_sign_assocs %>% rbind(fetched_df)
}
rnaseq_studies_sign_assocs$quant_method="ge"
```

```{r}
microarray_study_names <- c("CEDAR", "Fairfax_2014", "Kasela_2017", "Naranbhai_2015", "Fairfax_2012")
microarray_studies_sign_assocs <- data.frame()
for(study_name in microarray_study_names){
  print(study_name)
  ge_study_query_gwas_lead <- paste0("https://www.ebi.ac.uk/eqtl/api/associations?variant_id=rs4239702&p_upper=0.0001&quant_method=microarray&study=", study_name)
  fetched_df <- fetch_from_eqtl_cat_API(ge_study_query_gwas_lead)
  microarray_studies_sign_assocs <- microarray_studies_sign_assocs %>% rbind(fetched_df)
}
microarray_studies_sign_assocs$quant_method="microarray"
all_studies_sign_assocs <- rnaseq_studies_sign_assocs %>% rbind(microarray_studies_sign_assocs)
```

```{r}
check_colocalisation <- function(sign_assoc) {
  message(sign_assoc["qtl_group"])
  fetch_region_for_study_gene_link <- paste0("http://www.ebi.ac.uk/eqtl/api/chromosomes/",sign_assoc["chromosome"],
                                          "/associations?paginate=False&study=",sign_assoc["study_id"],
                                          "&qtl_group=",sign_assoc["qtl_group"],
                                          "&molecular_trait_id=",sign_assoc["molecular_trait_id"],
                                          "&quant_method=", sign_assoc["quant_method"],
                                          "&bp_lower=45980000&bp_upper=46200000&size=10000")
  
  message("Fetching: ", fetch_region_for_study_gene_link)
  region_for_study_gene_data <- fetch_from_eqtl_cat_API(fetch_region_for_study_gene_link)

  region_for_study_gene_data <- region_for_study_gene_data %>%
    dplyr::group_by(position) %>%
    dplyr::mutate(alt_allele_count = length(alt)) %>%
    dplyr::filter(alt_allele_count == 1)
  
  return(region_for_study_gene_data)
}
all_coloc_data <- list()
all_coloc_data$eqtl_assocs_in_region <- apply(all_studies_sign_assocs, 1, check_colocalisation)
```

```{r}
get_colocs <- function(eqtl_data_for_region, gwas_data_for_trait) {
  shared_positions = intersect(eqtl_data_for_region$position, gwas_data_for_trait$base_pair_location)

  eqtl_shared = dplyr::filter(eqtl_data_for_region, position %in% shared_positions) %>% 
    dplyr::mutate(variant_id = as.character(position))
  gwas_shared = dplyr::filter(gwas_data_for_trait, base_pair_location %in% shared_positions) %>% 
    dplyr::mutate(variant_id = as.character(base_pair_location))
  
  eQTL_dataset = list(pvalues = eqtl_shared$pvalue, 
                      N = 84, #The sample size of the eQTL dataset was 84
                      MAF = eqtl_shared$maf, 
                      type = "quant", 
                      beta = eqtl_shared$beta,
                      snp = eqtl_shared$variant_id)
  gwas_dataset = list(beta = gwas_shared$log_OR, #If log_OR column is full of NAs then use beta column instead
                      varbeta = gwas_shared$se^2, 
                      type = "cc", 
                      snp = gwas_shared$variant_id,
                      s = 0.5, #This is acutally not used, because we already specified varbeta above.
                      MAF = eqtl_shared$maf)
  
  coloc_res = coloc::coloc.abf(dataset1 = eQTL_dataset, dataset2 = gwas_dataset,p1 = 1e-4, p2 = 1e-4, p12 = 1e-5)
  return(coloc_res$summary)
}

all_coloc_data$colocs <- sapply(all_coloc_data$eqtl_assocs_in_region, get_colocs, gwas_data_for_trait=gwas_data)
```

Transfer colocalisation efficient to associations matrix
```{r}
for (index in 1:(all_coloc_data$eqtl_assocs_in_region %>% length())) {
  all_coloc_data$eqtl_assocs_in_region[[index]]$coloc_PP4 <- all_coloc_data$colocs[6,index]
}
```



```{r}
save_ggplots <- function(plot, path = ".", filename = "unnamed_plot", height = 15, width = 15){
  ggsave(plot = plot,
       filename = paste0(filename, ".eps"), 
       path = path,
       device = "eps", 
       height = height, 
       width = width,
       units = "cm",
       dpi = 300)

ggsave(plot = plot,
       filename = paste0(filename, ".png"), 
       path = path,
       device = "png", 
       height = height, 
       width = width,
       units = "cm",
       dpi = 300)

ggsave(plot = plot,
       filename = paste0(filename, ".pdf"), 
       path = path,
       device = "pdf", 
       height = height, 
       width = width,
       units = "cm",
       dpi = 300)
}
```

```{r}
get_plots <- function(eqtl_data_for_region, gwas_data_for_trait) {
  shared_positions = intersect(eqtl_data_for_region$position, gwas_data_for_trait$base_pair_location)

  eqtl_shared = dplyr::filter(eqtl_data_for_region, position %in% shared_positions) %>% 
    dplyr::mutate(variant_id = as.character(position))
  gwas_shared = dplyr::filter(gwas_data_for_trait, base_pair_location %in% shared_positions) %>% 
    dplyr::mutate(variant_id = as.character(base_pair_location))
  
  gwas_shared$track <- "GWAS RA"
  eqtl_shared$track <- paste0("eQTL Catalogue\n", unique(eqtl_data_for_region$study_id), "_", unique(eqtl_data_for_region$qtl_group))
  merged_data_for_plot <- gwas_shared %>% dplyr::select(p_value, base_pair_location, track) %>% stats::setNames(c("pvalue", "position", "track"))
  merged_data_for_plot <- merged_data_for_plot %>% rbind(as.data.frame(eqtl_shared %>% dplyr::select(pvalue, position, track)))
  merged_data_for_plot$track <- factor(merged_data_for_plot$track, levels = c(unique(gwas_shared$track), unique(eqtl_shared$track)))
  
  region_coords = c(45980000,46200000)
  plot_base_all = ggplot(merged_data_for_plot, aes_(x = ~position, y = ~-log(pvalue, 10))) + geom_blank() 
  
  plot_gwas_eqtl = plot_base_all + 
      # facet_grid(track_id ~ .) +
      geom_point(color = "deepskyblue4") + 
      ggtitle(paste0("PP4: ", round(unique(eqtl_data_for_region$coloc_PP4), digits = 3))) +
      theme_light() + 
      ylab(expression(paste("-",log[10], " p-value"))) +
      scale_x_continuous(limits = region_coords, expand = c(0,0)) + 
      facet_grid(track ~ ., scales = "free_y")+
      theme(plot.margin=unit(c(0.1,1,0.1,1),"line"),
          axis.text.x = element_blank(),
          axis.title.x = element_blank(), 
          axis.ticks.x = element_blank(),
          legend.position="none",
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          strip.text.y = element_text(colour = "grey10"),
          strip.background = element_rect(fill = "grey85"))
  
  plot_filename <- paste0(unique(gwas_shared$track), "_",unique(eqtl_shared$track), "_manhattan")
  message("Saving: ", plot_filename)
  save_ggplots(plot = plot_gwas_eqtl, path = "../paper_plots/eQTL_api_usecase/", filename = plot_filename, height = 12, width = 15)
  return(plot_gwas_eqtl)
}

all_coloc_data$manhattan_plots <- sapply(all_coloc_data$eqtl_assocs_in_region, get_plots, gwas_data_for_trait=gwas_data)
write_rds(all_coloc_data, "all_coloc_data.rds")
```

```{r}
# get_r2_for_lead_vars <- function(lead_var_eqtl, lead_var_gwas) {
#     out <- tryCatch(
#         {
#           LDpair(var1 = lead_var_gwas$variant_id, var2 = lead_var_eqtl$rsid, token = "here_goes_your_token")
#           message("Calculated R2 for:", lead_var_eqtl$rsid, " and ", lead_var_gwas$variant_id)
#         },
#         error=function(cond) {
#             message("Problem occured with variant: ", lead_var_eqtl$rsid, " from the study: ", lead_var_eqtl$study_id)
#             message("Here's the original error message:")
#             message(cond)
#             # Choose a return value in case of error
#             return(NA)
#         },
#         warning=function(cond) {
#             message(paste("variant caused a warning: ", lead_var_eqtl$rsid, " from the study: ", lead_var_eqtl$study_id))
#             message("Here's the original warning message:")
#             message(cond)
#             # Choose a return value in case of warning
#             return(NULL)
#         }
#     )    
#     return(out)
# }
```


```{r}
library("LDlinkR")

# calc_r2_for_leads <- function(eqtl_assocs_in_region, lead_var_gwas){
#   lead_var_eqtl = dplyr::arrange(eqtl_assocs_in_region, pvalue)[1,]
#   return(get_r2_for_lead_vars(lead_var_gwas = lead_var_gwas, lead_var_eqtl = lead_var_eqtl))
# }
# r2_of_leads <- sapply(all_coloc_data$eqtl_assocs_in_region, FUN = calc_r2_for_leads, lead_var_gwas = lead_var_gwas)
HipSci_assocs <- all_coloc_data$eqtl_assocs_in_region[[2]]
lead_var_eqtl = dplyr::arrange(HipSci_assocs, pvalue)[1,]
r2_hipsci_lead_var_vs_rs4239702 <- LDpair(var1 = lead_var_gwas$variant_id, var2 = lead_var_eqtl$rsid, token = "here_goes_your_token")
r2_hipsci_lead_var_vs_rs4239702
```


















